File: kakao/level1/실패율.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <iostream>
using namespace std;
vector<int> solution(int N, vector<int> stages) {
    vector<int> answer;
    vector<int> failed;
    vector<double> failrates;
    int clear = 0;
    answer.assign(N,0);
    failrates.assign(N,0);
    failed.assign(N,0);
    for(int i : stages){
        if((i-1)==N)
            clear++;
        else
            failed[i-1]++;
    }
    for(int i = N-1;i>=0;i--){
        answer[i] = i+1;
        clear += failed[i];
        if(clear != 0) failrates[i] = (double)failed[i] / clear;
    }
    stable_sort(answer.begin(),answer.end(),[&failrates](int i, int j) -> bool{return failrates[i-1]>failrates[j-1];});
    return answer;
}

File: 이것이 코딩테스트다/정렬/실패율.py
def solution(N, stages):
    fails, fail_rates = [0] * (N+1), [0] * N
    for stage in stages:
        fails[stage-1] += 1
    challengers = fails[N]
    for i in range(N - 1, -1, -1):
        challengers += fails[i]
        if challengers > 0:
            fail_rates[i] = fails[i] / challengers
    answer = list(map(lambda x:x[0]+1, sorted(enumerate(fail_rates),key=lambda x:(x[1]*-1,x[0]))))
    return answer

print(solution(15,[1,2,3,4,5,0,10]))